finished fetch all code templates feature

jerryy2021 opened the pull request:
**Summary**

An visitor can now fetch all code templates filtered by code content, title, and/or tags.

Note that this currently doesn't sort by relevance (the matched amount); this could be a future feature to implement.
**Review Notes**
Thanks to @Ram-Raghav-S , I implemented this feature by modifying his original code for get-comments.

**Testing**

**Test authenticated fetch:**
1. Login via login API
2. Save code template via its `POST save-code-template`
3. Save another code template via its sub-request `POST save-code-template-1`
4. Fetch using `GET fetch-all-code-templates` (should see two, or more results if existing templates also use given tags)
5. Fetch filtered using `GET filter-code-templates` (should see one less result than step 4)

**Test un-authenticated fetch:**
after running test for authenticated, do
1. Fetch using `GET visitor-fetch-all` API
Should see at least the same results as step 5 of previous test (provided the auth'ed user has userId 1).

Issues to be closed by this PR: 
- https://github.com/Scriptorium-CSC309/web/issues/25
- https://github.com/Scriptorium-CSC309/web/issues/22

Ram-Raghav-S reviewed (changes requested):
Clean code, although it needs a few changes

Fetch code templates with empty params gives no results

![image](https://github.com/user-attachments/assets/01391e43-949f-4fae-a1ea-60b42c309b13)

This is probably because even though tags, title, and code are marked as optional in the 
`Joi` schema, there is no check for if they exist before filtering on them.

Also, the issue with input validation could use some fixing.

Ram-Raghav-S commented on src/code-template/fetch-all-code-templates.ts:
```diff
@@ -0,0 +1,107 @@
+    .min(MIN_PAGE_SIZE)
+    .max(MAX_PAGE_SIZE)
+    .default(PAGE_SIZE),
+  tags: Joi.alternatives(
+    Joi.string().max(MAX_CHARS_TITLE_DESCRIPTION),
+    Joi.array().items(Joi.string().max(MAX_CHARS_TITLE_DESCRIPTION))
+  ).optional(),
+  title: Joi.string().max(MAX_CHARS_TITLE_DESCRIPTION).optional(),
+  code: Joi.string().max(MAX_CHARS_CONTENT).optional(),
```
while it is good to add this filter, the endpoint currently returns a 400 BAD REQUEST when for example tags are more than `MAX_CHAR_TITLE_DESCRIPTION` in length.

It is not really a bad request, maybe it would be when creating code-templates though.

You could simply leave out the check for max length here if you like.

Ram-Raghav-S commented:
Also, I know you referred to fetch comments, but fetch blogposts is closer to the fetch code template implementation. 
Looking there woul be helpful

jerryy2021 commented:
Oh I see. Great findings, thanks Ram! I've pushed my fixes. Also I just saw your where.OR approach which is fantastic. I saw yours after I fixed it using the approach GPT gave though, so I'd just leave it then. 

Also I added a new feature which is filter by userId and also allowing unauthenticated users to fetch content. I'll add documentation in the initial comment on how to test it.

Ram-Raghav-S commented:
Hey @jerryy2021 , I have rebased your branch and made a PR so that commits history makes sense

Here is the PR: https://github.com/Scriptorium-CSC309/web/pull/75

Could you see if I have correctly rebased your changes?

if yes, feel free to approve and that branch, the logic looks good

Ram-Raghav-S commented:
actually, I just noticed, why do you have the /public endpoint?

jerryy2021 commented:
Hi Ram, yeah I agree we probably don't need the protected endpoint. But according to this user story: 

https://github.com/Scriptorium-CSC309/web/issues/22 
> Make sure that only authenticated users can access this (fetch user code templates). 

The /public endpoint is used to implement this feature:
https://github.com/Scriptorium-CSC309/web/issues/25
> Allow vistors (not necessarily authenticated actors) to perform the above requirement using an API endpoint

So yeah, just for the TA testing's sake, I implemented both. We can decide to refactor it though. Also, should I close this PR without merging since we should merge the rebased PR instead?
